src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -244,6 +248,99 @@ public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions
             }
         }
 
+        internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+#if NETCOREAPP
+            bool success = TryParseEnumCore(ref reader, options, out T value);
+#else
+            bool success = TryParseEnumCore(reader.GetString(), options, out T value);
+#endif
+
+            if (!success)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            return value;
+        }
+
+        internal override void WriteAsPropertyNameCore(Utf8JsonWriter writer, T value, JsonSerializerOptions options, bool isWritingExtensionDataProperty)
+        {
+            ulong key = ConvertToUInt64(value);
+            string original = value.ToString();
+
+            if (IsValidIdentifier(original))
+            {
+                if (options.DictionaryKeyPolicy != null)
+                {
+                    original = options.DictionaryKeyPolicy.ConvertName(original);
+
+                    if (original == null)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_NamingPolicyReturnNull(options.DictionaryKeyPolicy);
+                    }
+
+                    writer.WritePropertyName(original);
+                    return;
+                }
+                else
+                {
+                    // We might be dealing with a combination of flag constants since all constant values were
+                    // likely cached during warm - up (assuming the number of constants <= NameCacheSizeSoftLimit).
+
+                    JavaScriptEncoder? encoder = options.Encoder;
+
+                    if (_nameCacheForWriting.Count < NameCacheSizeSoftLimit)
+                    {
+                        JsonEncodedText formatted = JsonEncodedText.Encode(original, encoder);
+
+                        writer.WritePropertyName(formatted);
+
+                        _nameCacheForWriting.TryAdd(key, formatted);
+                    }
+                    else
+                    {
+                        // We also do not create a JsonEncodedText instance here because passing the string
+                        // directly to the writer is cheaper than creating one and not caching it for reuse.
+                        writer.WritePropertyName(original);
+                    }
+
+                    return;
+                }
+            }
+
+            switch (s_enumTypeCode)
```

Comment:
Nit: there's a lot of overlap with what the Write method is doing -- could be extracted to a helper `WriteCore` method.